src/Toml/Parser.hs

```diff
@@ -106,18 +109,35 @@ boolP = False <$ text "false"
 -- dateTimeP = error "Not implemented!"
 
 arrayP :: Parser [UValue]
-arrayP = lexeme $ between (char '[' *> space) (char ']') (valueP `sepBy` spComma)
+arrayP = lexeme $ between (char '[' *> sc) (char ']') elements
   where
+    elements :: Parser [UValue]
+    elements = (choice (fmap (\p -> p `sepEndBy1` spComma) ps) <|> pure []) <* skipMany (text ",")
```

Comment:
You don't need to implement parser in a such way to parse elements of single type. This is a part of typechecking:

https://github.com/kowainik/tomland/blob/9d44db6f56088c38d76a443b74defbd5ebd5245b/src/Toml/Parser.hs#L128-L131

And it was done in a such way to provide better error messages in future. With catching this on parser stage you can't give good error message without making parser scary. So it's okay if array parse can parse `"[true, 42]"`.

Reply:
The main reason for creating these extra functions (`uBoolP`, `uFloatP`, `uIntP`, `uStringP`, `uArrayP`), is to marke `arrayP` parse elements of a single type.

If it's okay for `arrayP` to parse elements of different types, then adding those functions would unnecessary, and I can go back to what we have in `master`:
https://github.com/kowainik/tomland/blob/9d44db6f56088c38d76a443b74defbd5ebd5245b/src/Toml/Parser.hs#L114-L120

Reply:
@ghallak It's completely okay to allow arrays of different elements during `arrayP` parser. Because same-type validation is performed later in parsing.